src/main/java/org/numenta/nupic/encoders/RandomDistributedScalarEncoder.java

```diff
@@ -0,0 +1,676 @@
+	 * @throws IllegalStateException
+	 */
+	public boolean newRepresentationOK(List<Integer> newRep, int newIndex)
+			 {
+		if (newRep.size() != getW())
+			return false;
+		if (newIndex < getMinIndex() - 1 || newIndex > getMaxIndex() + 1)
+			throw new IllegalStateException(
+					"newIndex must be within one of existing indices");
+
+		// A binary representation of newRep. We will use this to test
+		// containment
+		boolean[] newRepBinary = new boolean[getN()];
+		Arrays.fill(newRepBinary, false);
+		for (int index : newRep)
+			newRepBinary[index] = true;
+
+		// Midpoint
+		int midIdx = getMaxBuckets() / 2;
+
+		// Start by checking the overlap at minIndex
+		int runningOverlap = countOverlap(bucketMap.get(getMinIndex()), newRep);
+		if (!overlapOK(getMinIndex(), newIndex, runningOverlap))
+			return false;
+
+		// Compute running overlaps all the way to the midpoint
+		for (int i = getMinIndex() + 1; i < midIdx + 1; i++) {
+			// This is the bit that is going to change
+			int newBit = (i - 1) % getW();
+
+			// Update our running overlap
+			if (newRepBinary[bucketMap.get(i - 1).get(newBit)])
+				runningOverlap--;
+			if (newRepBinary[bucketMap.get(i).get(newBit)])
+				runningOverlap++;
+
+			// Verify our rules
+			if (!overlapOK(i, newIndex, runningOverlap))
+				return false;
+		}
+
+		// At this point, runningOverlap contains the overlap for midIdx
+		// Compute running overlaps all the way to maxIndex
+		for (int i = midIdx + 1; i <= getMaxIndex(); i++) {
+			int newBit = i % getW();
+
+			// Update our running overlap
+			if (newRepBinary[bucketMap.get(i - 1).get(newBit)])
+				runningOverlap--;
+			if (newRepBinary[bucketMap.get(i).get(newBit)])
+				runningOverlap++;
+
+			// Verify our rules
+			if (!overlapOK(i, newIndex, runningOverlap))
+				return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Return the overlap between two representations. rep1 and rep2 are lists
+	 * of non-zero indices.
+	 * 
+	 * @param rep1
+	 * @param rep2
+	 * @return
+	 */
+	public int countOverlap(List<Integer> rep1, List<Integer> rep2) {
+		int overlap = 0;
+		for (int index : rep1) {
+			for (int index2 : rep2)
+				if (index == index2)
+					overlap++;
+		}
+		return overlap;
+	}
+
+	/**
+	 * Return the overlap between two representations. rep1 and rep2 are arrays
+	 * of non-zero indices.
+	 * 
+	 * @param rep1
+	 * @param rep2
+	 * @return
+	 */
+	public int countOverlap(int[] rep1, int[] rep2) {
+		int overlap = 0;
+		for (int index : rep1) {
+			for (int index2 : rep2)
+				if (index == index2)
+					overlap++;
+		}
+		return overlap;
+	}
+
+	/**
+	 * Return True if the given overlap between bucket indices i and j are
+	 * acceptable.
+	 * 
+	 * @param i
+	 * @param j
+	 * @param overlap
+	 * @return
+	 */
+	public boolean overlapOK(int i, int j, int overlap) {
+		if (Math.abs(i - j) < getW())
+		{
+			if((overlap == (getW() - Math.abs(i - j))))
+				return true;
+			else
+				return false;
+		}
+		else
+		{
+			if(overlap <= getMaxOverlap())
```

Comment:
I don't mean to propose a change to the logic, just to the implementation:

    if (Math.abs(i - j) < w) {
        return overlap == w - Math.abs(i - j);
    } else {
        return overlap <= maxOverlap;
    }

It just reads cleaner to me as "return the truth of X" rather than, "is X true? return true, otherwise return false"